lockfile_generator/src/yarn.rs

```diff
@@ -13,7 +15,38 @@ impl Generator for Yarn {
         let project_path = manifest_path
             .parent()
             .ok_or_else(|| Error::InvalidManifest(manifest_path.to_path_buf()))?;
-        Ok(project_path.join("yarn.lock"))
+
+        let output = Command::new("yarn")
+            .current_dir(project_path)
+            .args(["workspaces", "list", "--json"])
+            .output()?;
+
+        // Ensure command was successful.
+        if !output.status.success() {
+            let stderr = String::from_utf8_lossy(&output.stderr);
+            return Err(Error::NonZeroExit(output.status.code(), stderr.into()));
+        }
+
+        // Convert output to actual valid json.
+        let stdout = String::from_utf8(output.stdout).map_err(Error::InvalidUtf8)?;
+        let workspaces_json = format!("[{}]", stdout.trim_end().replace('\n', ","));
+
+        // Parse workspace list.
+        let mut workspaces: Vec<Workspace> = serde_json::from_str(&workspaces_json)?;
+
+        // Sort by longest location path, to prefer longer matches.
+        workspaces.sort_by(|a, b| b.location.len().cmp(&a.location.len()));
+
+        // Find workspace root by stripping the first matching location path.
+        let project_str = project_path
+            .to_str()
+            .ok_or_else(|| Error::InvalidManifest(manifest_path.to_path_buf()))?;
+        let workspace_root = workspaces
+            .into_iter()
+            .find_map(|project| project_str.strip_suffix(&project.location))
```

Comment:
A contrived example:

```sh-session
> pwd
/Users/kyle/tmp-js/foo

> yarn workspaces list --json
{"location":".","name":"tmp-js"}
{"location":"foo","name":"foo"}
{"location":"bar","name":"bar"}
{"location":"js/foo","name":"js-foo"}
```

This would incorrectly pick `/Users/kyle/tmp-/yarn.lock` as the lockfile location.

I don't think `yarn workspaces list --json` actually gives enough information to correctly find the root workspace...

Reply:
Right that's why I originally used `ends_with` on the path, which I promptly forgot about apparently.

Would you suggest removing yarn completely then? It's the best I found.

Reply:
It's not bad enough to drop support for yarn... This still improves the status quo. We just have to accept the possibility of this annoying edge case.

Using `ends_with` to ensure that the break happens between path components would certainly be a welcome improvement.

Reply:
For another approach, take a look at the [`find-yarn-workspace-root` project](https://github.com/square/find-yarn-workspace-root/tree/master), which itself [adapted the algorithm directly from `yarn`](https://github.com/square/find-yarn-workspace-root/blob/master/index.js#L11), here:

https://github.com/yarnpkg/yarn/blob/ddf2f9ade211195372236c2f39a75b00fa18d4de/src/config.js#L612

Granted, this reference is for the `yarn` v1.x branch and the same code could not be found [in the current v3.x branch](https://github.com/yarnpkg/berry), but the algorithm appears to still be valid.

Reply:
Doesn't look too bad, I can probably adapt that. Though I hate having to rely on an ever-changing implementation like that.